### ISSUE TITLE
NullPointerException in ConstraintsValidator.java:36 when sending UPDATE query

### ISSUE THREAD

**Issue by cweiske** [crate/crate#4594]
**CrateDB version**:  0.56.3
**JVM version**: 1.8.0_92-internal
**OS version / environment description**: `crate/crate` docker image `81aa2e52c2cf`

**Problem description**:

We get a `NullPointerException` when sending a SQL `UPDATE` statement to crate:

```
HTTP/1.1 500 Internal Server Error\r\n

{
    "error": {
        "message": "SQLActionException[NullPointerException in io.crate.analyze.ConstraintsValidator.validate(ConstraintsValidator.java:36)]",
        "code": 5000
    }
}
```

Request:
```
{
    "stmt": "update production_users set username = ?, email = ?, salutation = ?, firstname = ?, lastname = ?, role = ?, businesstype = ?, active = ?, copycontactemail = ?, street = ?, housenumber = ?, city = ?, zip = ?, country = ?, phone = ?, mobilephone = ?, fax = ?, website = ?, text = ?, image = ?, profile = ?, imprint = ?, meta = ?, contract = ?, migration = ?, customer_id = ?, last_editor_id = ?, creator_id = ?, deleted_at = ?, producer = ?, hasfullprofile = ?, hasbpalogo = ?, hascontract = ?, haspubliccontacts = ?, hasdatafacts = ?, hasprofiledata = ?, brokername = ?, locations = ?, contacts = ?, updated_at = ? where id = ?",
    "args": [
        "",
        "test@example.org",
        "",
        "",
        "",
        "producer",
        "commercial",
        1,
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        "",
        {
            "about_us": "",
            "benefits": ""
        },
        "{\"portrait\":\"\",\"logo\":\"\",\"companylogo\":\"\",\"sellogo\":\"\",\"polelogo\":\"\",\"company\":\"\"}",
        {
            "company_name": "",
            "company_type": "",
            "founding_year": "",
            "employee_count": 0,
            "languages": "[]",
            "memberships": "[]",
            "slogan": "",
            "description": "",
            "services": "[]",
            "servicestext": "",
            "facebook": "",
            "twitter": "",
            "googleplus": "",
            "linkedin": "",
            "xing": "",
            "image": "",
            "publicvisible": "",
            "phone": "",
            "fax": "",
            "email": "",
            "youtube": "",
            "keyvalue": ""
        },
        {
            "companyname": "",
            "street": "",
            "housenumber": "",
            "city": "",
            "zip": "",
            "country": "",
            "phone": "",
            "fax": "",
            "email": "test@example.org",
            "website": "",
            "representative": "",
            "inspectorate": "",
            "jobcham": "",
            "no_inspectorate": 0,
            "hrb": "",
            "no_hrb": 0,
            "vat_id": "",
            "no_taxid": 0,
            "additional_text": "",
            "disclaimer": ""
        },
        {
            "bpa": 0,
            "bpa_end": "0000-00-00",
            "premiumpresentation": 0,
            "comment": "",
            "software": ""
        },
        {
            "name": "",
            "max_objects": 0,
            "start": null,
            "end": null,
            "bill_salutation": "",
            "bill_firstname": "",
            "bill_lastname": "",
            "bill_company": "",
            "bill_street": "",
            "bill_housenumber": "",
            "bill_city": "",
            "bill_zip": "",
            "bill_country": "",
            "bonitaet": "",
            "coc": "",
            "ioc": "",
            "toc": "",
            "eoc": "",
            "runtimetype": "",
            "minimum_term": "",
            "renewal_interval": "",
            "category": "",
            "type": "",
            "fixed_price": "",
            "backoffice_price": "",
            "benefit_end": "",
            "benefit_reason": "",
            "price_group": ""
        },
        {
            "shopid": "",
            "personid": ""
        },
        null,
        null,
        null,
        null,
        {
            "dealerlist": "",
            "sectors": ""
        },
        0,
        0,
        0,
        0,
        0,
        0,
        "",
        [],
        [],
        "1481708697",
        23563
    ]
}
```

**Steps to reproduce**:

Schema:
```
CREATE TABLE IF NOT EXISTS "doc"."production_users" (
   "active" INTEGER,
   "brokername" STRING,
   "businesstype" STRING,
   "city" STRING,
   "contacts" ARRAY(STRING),
   "contract" OBJECT (DYNAMIC),
   "copycontactemail" STRING,
   "country" STRING,
   "created_at" TIMESTAMP,
   "customer_id" INTEGER,
   "deleted_at" TIMESTAMP,
   "email" STRING,
   "fax" STRING,
   "firstname" STRING,
   "hasbpalogo" SHORT,
   "hascontract" SHORT,
   "hasdatafacts" SHORT,
   "hasfullprofile" SHORT,
   "hasprofiledata" SHORT,
   "haspubliccontacts" SHORT,
   "housenumber" STRING,
   "id" LONG,
   "image" STRING INDEX OFF,
   "imprint" OBJECT (DYNAMIC),
   "lastname" STRING,
   "locations" ARRAY(STRING),
   "meta" OBJECT (DYNAMIC) AS (
      "bpa" INTEGER,
      "premiumpresentation" INTEGER
   ),
   "migration" OBJECT (DYNAMIC),
   "mobilephone" STRING,
   "phone" STRING,
   "position" GEO_POINT,
   "profile" OBJECT (DYNAMIC) AS (
      "company_name" STRING
   ),
   "role" STRING,
   "salutation" STRING,
   "street" STRING,
   "text" OBJECT (DYNAMIC),
   "updated_at" TIMESTAMP,
   "username" STRING,
   "website" STRING,
   "zip" STRING,
   PRIMARY KEY ("id")
)
```

**Comment by celaus** [comment]
Hey, thanks for submitting, I'll take a look!

**Comment by chaudum** [comment]
hi @cweiske 

the problem was that CrateDB tried to validate a column (`producer`) that did not exist before the update against constraints (in this case `not null` because the value is `NULL`). since the column is only created later, the NPE occurred.
this issue is fixed on master on branch 1.0 and will be available with the next hotfix release, see #4614

**Comment by cweiske** [comment]
Thank you for fixing this. I already feared that the data in the report were not enough, but that's all I had from the wireshark dump of the app <-> crate traffic.

**Comment by matriv** [comment]
Hello @cweiske,

I'm closing this issue. The fix will be available in 1.0.2 release.
Thanks again for reporting!
